### ISSUE TITLE
When `ListenableEditingState` is provided `initialState` with a valid composing range, `android.view.inputmethod.BaseInputConnection.setComposingRegion(int, int)` has NPE

### ISSUE THREAD

**Issue by fzyzcjy** [flutter/flutter#96570]
Thanks for the framework! But I see a bug:

## Steps to Reproduce

Reproduced in production environments many times, so no easy steps to do here. But plz see the following logical analysis - it is very easy to understand.

## Stack trace

Please see analysis below to understand why it happens.

```
PlatformException(error, Attempt to invoke virtual method 'boolean android.view.inputmethod.BaseInputConnection.setComposingRegion(int, int)' on a null object reference, null, java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.view.inputmethod.BaseInputConnection.setComposingRegion(int, int)' on a null object reference
	at io.flutter.plugin.editing.ListenableEditingState.setComposingRange(SourceFile:1)
	at io.flutter.plugin.editing.ListenableEditingState.setEditingState(SourceFile:6)
	at io.flutter.plugin.editing.ListenableEditingState.<init>(SourceFile:7)
	at io.flutter.plugin.editing.TextInputPlugin.setTextInputClient(SourceFile:9)
	at io.flutter.plugin.editing.TextInputPlugin$1.setClient(SourceFile:1)
	at io.flutter.embedding.engine.systemchannels.TextInputChannel$1.onMethodCall(SourceFile:39)
	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(SourceFile:2)
	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(SourceFile:2)
	at io.flutter.embedding.engine.dart.DartMessenger.a(SourceFile:2)
	at io.flutter.embedding.engine.dart.DartMessenger.b(SourceFile)
	at i.a.a.b.b.a.run(SourceFile)
	at android.os.Handler.handleCallback(Handler.java:761)
	at android.os.Handler.dispatchMessage(Handler.java:98)
	at android.os.Looper.loop(Looper.java:156)
	at android.app.ActivityThread.main(ActivityThread.java:6623)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:942)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:832)
)
  libapp              0x95b3d9f6   JSONMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:164)
  libapp              0x951edd04   MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:167)
```

## Analysis

Remark: All analysis below refer to https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java

Provided that, `ListenableEditingState` is constructed with `initialState` having a valid composing range. Then the following analysis shows that it NPE.

Firstly, constructor runs:

https://github.com/flutter/engine/blob/29abb50f256a8156bb5ab36b435f9c18f2510ec9/shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java#L60-L75

Then in line 64, it calls `setEditingState`. The latter then calls `setComposingRange`, which looks like below:

https://github.com/flutter/engine/blob/29abb50f256a8156bb5ab36b435f9c18f2510ec9/shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java#L140-L146

Since the initialState provided has a valid composing, it runs:

https://github.com/flutter/engine/blob/29abb50f256a8156bb5ab36b435f9c18f2510ec9/shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java#L144

However, by naively looking at the `mDummyConnection` variable, we see it is not assigned a value until the constructor is going to be finished:

https://github.com/flutter/engine/blob/29abb50f256a8156bb5ab36b435f9c18f2510ec9/shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java#L60-L75

In short: line 64 calls `mDummyConnection.somefunction()` while `mDummyConnection` is initialized in line 68.

```
no error
```

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 11.6.1 20G224 darwin-x64, locale zh-Hans-CN)
    • Flutter version 2.8.1 at /Users/tom/opt/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (4 周前), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn
```

</details>

**Comment by darshankawar** [comment]
@fzyzcjy 
Thanks for the detailed analysis and potential fix. Looking at the error:

`
PlatformException(error, Attempt to invoke virtual method 'boolean android.view.inputmethod.BaseInputConnection.setComposingRegion(int, int)' on a null object reference, null, java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.view.inputmethod.BaseInputConnection.setComposingRegion(int, int)' on a null object reference`

Seems to be pointing to `BaseInputConnection` class and further `setComposingRegion`, method which seems to getting called on null. Would you know the device on which this might be happening ? (e.g: Samsung or other make)

**Comment by fzyzcjy** [comment]
> Seems to be pointing to BaseInputConnection class and further setComposingRegion, method which seems to getting called on null

The core reason is described as above. `      mDummyConnection.setComposingRegion(composingStart, composingEnd);` throw error because `mDummyConnection` is null.

> Would you know the device on which this might be happening ? (e.g: Samsung or other make)

Not specific to one device kind. It is caused by some of my code using TextField. But those logic are quite complex and I cannot write down a reproducible example without giving you tons of code.

**Comment by fzyzcjy** [comment]
Indeed, just by logic thinking: When the `ListenableEditingState` is constructed with a valid composing range as its constructor arguments, this bug **must happen**. So this is a bug and should be fixed imho.

**Comment by darshankawar** [comment]
Thanks for the update. I am labeling it for further input from the team on this.

**Comment by fzyzcjy** [comment]
@darshankawar You are welcome

**Comment by kobi666** [comment]
is there a workaround of some sort?

**Comment by kobi666** [comment]
can you at least tell us which version this bug was introduced at so we can know to which version to downgrade?

**Comment by fzyzcjy** [comment]
@kobi666 

The bug occurs when we test it recently, so I guess it may be related to Flutter 2.8 (I am using 2.8.1 indeed). However, it may also be possible that it is introduced earlier, but we did not find it because the code logic was slightly different or we did not manually test this strange situation.

**Comment by kobi666** [comment]
> @kobi666
> 
> The bug occurs when we test it recently, so I guess it may be related to Flutter 2.8 (I am using 2.8.1 indeed). However, it may also be possible that it is introduced earlier, but we did not find it because the code logic was slightly different or we did not manually test this strange situation.
> 
> Indeed, by logical reasoning, it should exist for a long long time.
> 
> I also hope someone in Flutter team can give some suggestions, or fix it in the next release. I am willing to make a PR.

thanks,
I also think it is very very recent as i didn't work on a textfield in my proj for two weeks (which worked fine) and then suddenly this. (i probably did a flutter upg or something)

**Comment by kobi666** [comment]
@darshankawar 
Hi,
is there any new update on this?
I'm kind of worried to downgrade my project to solve this (tho if I will, OC the thread will be updated with the results)

**Comment by FabioPaganoAPSystem** [comment]
I've noticed that the problem arises only if the textfield has some text in it, so in my case a temporary workaround was to empty the field after having pressed the button "search" (because the field is used to search on a database based on the entered text). I've also noticed that the problem appears only if I have reloaded the screen, because the first time it doesn't appear (also if I tap on the textfield with text in it) but appears only if I click "search" and then the screen is refreshed with the fetched data, from that moment on if I tap on the search field (still filled with the previous search text) the problem arises. The problem is only in Android (tested good on iOS and Web), in flutter version 2.0.1 didn't appear. Thank you.

**Comment by chinmaygarde** [comment]
cc @blasten We tried to follow along what could be happening in code in triage and it looks like there is a method call in the constructor on a potentially incompletely initialized object.

**Comment by chinmaygarde** [comment]
cc @LongCatIsLooong @blasten The analysis in the original post on how this could be happening is accurate. @fzyzcjy Can you submit a patch for this?

**Comment by kobi666** [comment]
@chinmaygarde @darshankawar 
Hey,
I just confirmed that this is not related to 2.8.1 flutter version.
I migrated my project to a new one at 2.5.3, and the issue is still persistent.

Can we get some kind of workaround please?

also, @FabioPaganoAPSystem 
the issue for me at least occurs even when the textfield is empty.

doctor output:
flutter doctor -v
[√] Flutter (Channel stable, 2.5.3, on Microsoft Windows [Version 10.0.18363.1316], locale en-IL)
    • Flutter version 2.5.3 at D:\flutter\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 18116933e7 (3 months ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4

**Comment by fzyzcjy** [comment]
@chinmaygarde Sure

**Comment by fzyzcjy** [comment]
https://github.com/flutter/engine/pull/30916

**Comment by kobi666** [comment]
further testing,
When I create a new project from scratch (VS code example project), the issue (that is, the reload of the page and textfield not working) does not occur.
I still get the messages in the debug console tho:

W/IInputConnectionWrapper(23144): beginBatchEdit on inactive InputConnection
W/IInputConnectionWrapper(23144): getTextBeforeCursor on inactive InputConnection
W/IInputConnectionWrapper(23144): getTextAfterCursor on inactive InputConnection
W/IInputConnectionWrapper(23144): getSelectedText on inactive InputConnection
W/IInputConnectionWrapper(23144): endBatchEdit on inactive InputConnection
W/IInputConnectionWrapper(23144): beginBatchEdit on inactive InputConnection
W/IInputConnectionWrapper(23144): getTextBeforeCursor on inactive InputConnection
W/IInputConnectionWrapper(23144): getTextAfterCursor on inactive InputConnection
W/IInputConnectionWrapper(23144): getSelectedText on inactive InputConnection
W/IInputConnectionWrapper(23144): endBatchEdit on inactive InputConnection

i'm now re adding my project manually to see where the issue is triggered.
